mage_ai/streaming/sources/google_cloud_pubsub.py

```diff
@@ -0,0 +1,127 @@
+                return True
+        return False
+
+    def _create_subscription(
+            self, project_id: str, topic_id: str, subscription_id: str) -> None:
+        """Create a new pull subscription on the given topic."""
+        # [START pubsub_create_pull_subscription]
+        if self._exist_subscription(project_id):
+            self._print(f"Subscription already exists: {self.subscription_path}")
+            return
+
+        publisher = pubsub_v1.PublisherClient()
+        topic_path = publisher.topic_path(project_id, topic_id)
+        subscription = self.subscriber_client.create_subscription(
+            request={"name": self.subscription_path, "topic": topic_path}
+        )
+        self._print(f"Subscription created: {subscription}")
+        # [END pubsub_create_pull_subscription]
+
+    def init_client(self) -> None:
+        if self.config.pubsub_emulator_host is not None:
+            os.environ["PUBSUB_EMULATOR_HOST"] = self.config.pubsub_emulator_host
+        if self.config.google_application_credentials is not None:
+            os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = \
+                self.config.google_application_credentials
+        self.subscriber_client = pubsub_v1.SubscriberClient()
+        self.subscription_path = self.subscriber_client.subscription_path(
+            self.config.project_id, self.config.subscription_id)
+        self._create_subscription(
+            self.config.project_id,
+            self.config.topic_id,
+            self.config.subscription_id)
+
+    def read(self, handler: Callable) -> None:
+        self._print("Start consuming messages.")
+
+        def callback(
+                received_message: pubsub_v1.subscriber.message.Message) -> None:
+            self._print(f"Received: {received_message}.")
+            handler(dict(data=received_message.message.data.decode()))
+            self._print(f"Handled: {received_message.message.data}.")
+            received_message.ack()
+
+        with self.subscriber_client:
+            streaming_pull_future = self.subscriber_client.subscribe(
+                self.subscription_path, callback=callback)
+
+            try:
+                # When `timeout` is not set, result() will block indefinitely,
+                # unless an exception is encountered first.
+                self._print("Start receiving message with timeout: {self.config.timeout}")
+                streaming_pull_future.result(timeout=self.config.timeout)
+            except TimeoutError:
+                streaming_pull_future.cancel()  # Trigger the shutdown.
+                streaming_pull_future.result()  # Block until the shutdown is complete.
+
+    def batch_read(self, handler: Callable) -> None:
+        self._print("Start consuming messages.")
+        if self.config.batch_size > 0:
+            batch_size = self.config.batch_size
+        else:
+            batch_size = DEFAULT_BATCH_SIZE
+
+        with self.subscriber_client:
```

Comment:
could you wrap the code with the "while True" loop so that it can continue pulling messages?

Reply:
Made the suggested change.